readme.md
LogTrace
========

[![Build Status](https://travis-ci.org/paul-wolf/logtrace.svg?branch=master)](https://travis-ci.org/paul-wolf/logtrace)

Aggregate messages to produce a log entry representing a single event or procedure. The purpose of this module is to easily asssociate log messages together that belong together.

```
import logging
from logtrace import LogTrace

logger = logging.getLogger(__name__)
trace = LogTrace(logger=logger)

trace.add("Let's get started")
...
trace.add("Later, something else happens")
...
trace.add("And finally...")

trace.emit()
```

You get a single log entry like this:

```
[05/Jan/2018 11:11:00] DEBUG [21, .30s] Let's get started; [65, .132s] Later, something else happens; [75, .330s] And finally...
```

Install
-------

	pip install logtrace

Note that this only suppports Python 3. Let me know if anyone wants support for Python 2. There are no dependencies outside the Python Standard Library for this module. 

Example
-------

Logs can be hard to read because you have cases where you log information as you go through a procedure. These log entries get scattered with all the other logs from other processes. You end up having to search for related entries possibly implanting identifying information in each one to tie them together. `LogTrace` fixes this problem by letting you collect logs and then output once. Take the example of a token authentication procedure where transient tokens are required to be authenticated. You want to record the following events:

* Check the HTTP header info with the token
* What table are we going to use to check the token?
* Did the token service authenticate the token?
* Is the token in a local cache?
* Successfully authenticated? 

The following records five separate instances where you would have called `logger.info()` with a line number and the time in seconds since constructing the `LogTrace` object `[<lineno>, <secs>s]`: 

```
[12:12:54] INFO [132, 0.0006s] auth header: [b'Token', b'2c59999137******************************']; [132, 0.0007s] authenticate key, model: <class 'tastypie.models.ApiKey'>; [132, 0.1057s] token renewal for API call confirmed; [132, 0.1078s] got key from token table: paul; [163, 0.1079s] Successfully authenticated
```

Details
-------

We respect logging levels. So, the overhead of using LogTrace is minimal if your log level is not effective. If your log level is `logging.INFO` and you call `logtrace.emit_debug()`, almost all overhead is avoided minus some function call overhead and one or two conditional expressions. 

What LogTrace is *not*: This is *not* a logging framework. LogTrace uses the standard Python `logging` module. All your configuration to `logging` is going to be used by LogTrace. All your handlers are going to act exactly as before. If you use a framework like Django, you use it just like you do now. No changes whatever are required to your logging configuration. 

We also provide other features like

* Easily generate a UUID for the logged event.

* Timing for each message since LogTrace was created.

* Frame information for each part message, like filename, function, lineno

* Any logging mechanism can be used, not just standard Python logging.

* Pass structured data (JSON).

We wanted to provide something that works in perfect harmony with the
existing Python logging module without unnecessary duplication of
features and no external dependencies (outside the PSL).

```
    LogTrace(logger=None,      # we'll emit output here
             delimiter="; ",   # delimiter between messages
             tag='',           # add a non-unique label 
             unique_id=False,  # create a uuid to identify the log?
	         verbosity='v'     # level of output for frame information
            )
```

* `logger`: the standard logger returned from `import logging; logger
  = logging.getLogger(__name__)`. You can create a `LogTrace()`
  without a logger in which case it creates with the value of `__name__`.

* `delimiter`: the character(s) used between messages

* `tag`: This is a convenience to tell LogTrace() to use hash+tag at
  the start of every entry after calling `.emit()` for ease of
  searching.

* `unique_id`: generate a uuid to associate with the final message output.

* `verbosity`: v, vv, vvv for three levels of verbosity when adding
  frame information

`LogTrace.get_uid()`: return the unique id. If one has not been set during construction of the LogTrace, a uuid is generated. Otherwise, it returns the existing one. 

`LogTrace.set_uid(uid)`: Set a unique id. This can be done by constructing `LogTrace()` with `unique_id=True`. This takes normally either a uuid or str argument.  

`LogTrace.add(msg, data, backup)`: Add a message to the list. This will get frame information for the call depending on the verbosity level. 

`LogTrace.emit_string()`: return a string that is the final log message.

`LogTrace.emit()`: call `logger.debug(message)` 

`LogTrace.emit_error()`: call `logger.error(message)`

`LogTrace.emit_info()`: call `logger.info(message)`

`LogTrace.emit_debug()`: call `logger.debug(message)`

`LogTrace.emit_warning()`: call `logger.warning(message)`

`LogTrace.emit_critical()`: call `logger.critical(message)`

When the `LogTrace` is created, `time.time()` is recorded. Whenever `LogTrace.add()` is called, the start time is subtracted from the current time when the message is added. The final message prints the number of seconds since creating.

You probably want to avoid including `LogTrace.add()` in loops. You also probably want to create it as a local, not a module-level variable. Pass it as a method argument rather than using a module level instance. If you do want to re-use a `LogTrace` and clear messages, you can call `LogTrace.clear()`. But be aware the uid might need to be reset depending on your application requirements.

Extra Data
----------

`LogTrace.add()` has an optional parameter `data` that takes a dictionary. We keep a dict in the object and `update()` it whenever the `data` parameter is used. This doesn't do anything within `LogTrace` itself other than maintain the `data` member variable. But you can accumulate data and later ship the data to a service like AWS S3 or whatever, like this:

    logger.info(trace.emit_string(), extra=trace.data)

This would be useful if you are using a logging handler that ships the `logging.LogRecord` as JSON to some service like a document oriented data store, Elasticsearch, etc.

You also could pass extra params via `.emit` call like
```
trace = LogTrace(logger=logger)
trace.emit('message', extra={'key': 'value'})
```

Testing
-------

	pip install pytest
	cd logtrace
 	pytest test.py --verbose

or

	python3 logtrace/test.py

Performance
-----------

`LogTrace()` appends to a list of strings everytime you call `add()`. But it firstly calls `inspect.getFrameInfo()` and builds the string with that information. When `emit()` is called, it concatenates all the strings in the list separated by `delimiter` and then calls `logger.info()` or whatever method is appropriate. If the effective level is not the current level for the method, then the list will be empty and it won't do the call to the `logger` method.

Acknowledgements
----------------

Thanks to

.. @metazet: https://github.com/metazet

For important fixes. 


CHANGES.txt
Changes 0.1.1
-------------

@metazet fixed delimiter bug and extra parameter in emit()



logtrace/test.py
# -*- coding: utf-8 -*-
import uuid
import unittest
import logging
from __init__ import LogTrace

logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s %(levelname)-8s %(message)s',
                    datefmt='%Y-%m-%d %H:%M:%S')

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
s = [98, 127, 133, 147, 170, 197, 201, 211, 255]


class TestLogTrace(unittest.TestCase):
    def test_logtrace_simple(self):
        trace = LogTrace(logger, unique_id=True, verbosity='v')
        trace.add("first message")
        trace.add("second message")
        print(trace.emit_string())
        trace.emit("finally")

    def test_logtrace_unique_id(self):
        trace = LogTrace(logger=logger, unique_id=True, level=logging.INFO, verbosity='vvv')
        trace.add("first message")
        trace.add("second message")
        print(trace.emit_string())

    def test_function(self):
        trace = LogTrace(logger=logger, level=logging.INFO)
        standard_deviation(s, trace=trace)
        print(trace.emit_string())

    def test_function_tag(self):
        trace = LogTrace(logger=logger, tag='STDDEV', level=logging.INFO, verbosity='vv')
        standard_deviation(s, trace=trace)
        print(trace.emit_string())

    def test_logtrace_uuid(self):
        trace = LogTrace(logger, unique_id=True, level=logging.INFO)
        trace.add("first message")
        trace.add("second message")
        print(trace.get_uid())
        trace.set_uid(uuid.uuid4())
        trace.set_uid(str(uuid.uuid4()))  # could be a string
        trace.emit("finally")

    def test_unicode(self):
        trace = LogTrace(logger)
        s = "ƀ Ɓ Ƃ ƃ Ƅ ƅ Ɔ Ƈ ƈ Ɖ Ɗ Ƌ ƌ ƍ Ǝ Ə Ɛ Ƒ ƒ Ɠ Ɣ ƕ Ɩ Ɨ Ƙ ƙ ƚ ƛ Ɯ Ɲ ƞ Ɵ Ơ ơ Ƣ ƣ Ƥ ƥ Ʀ Ƨ ƨ Ʃ ƪ ƫ Ƭ ƭ Ʈ Ư ư Ʊ Ʋ Ƴ ƴ Ƶ ƶ Ʒ Ƹ ƹ ƺ ƻ Ƽ ƽ ƾ ƿ ǀ ǁ ǂ ǃ Ǆ ǅ ǆ Ǉ ǈ ǉ Ǌ ǋ ǌ Ǎ ǎ Ǐ ǐ Ǒ ǒ Ǔ ǔ Ǖ ǖ Ǘ ǘ Ǚ ǚ Ǜ ǜ ǝ Ǟ ǟ Ǡ ǡ Ǣ ǣ Ǥ ǥ Ǧ ǧ Ǩ ǩ Ǫ ǫ Ǭ ǭ Ǯ ǯ ǰ Ǳ ǲ ǳ Ǵ ǵ Ǻ ǻ Ǽ ǽ Ǿ ǿ Ȁ ȁ Ȃ ȃ"
        trace.add(s)
        print(trace.emit_string())

    def test_extra(self):
        trace = LogTrace(logger)
        trace.emit('finish', extra={'key': 'value'})


def standard_deviation(lst, population=True, trace=None):
    """Calculates the standard deviation for a list of numbers.

    Just for testing LogTrace().

    """

    num_items = len(lst)
    trace.add('num_items={}'.format(num_items))

    mean = sum(lst) / num_items
    trace.add('mean={}'.format(mean))

    differences = [x - mean for x in lst]
    sq_differences = [d ** 2 for d in differences]
    ssd = sum(sq_differences)
    trace.add('ssd={}'.format(ssd))

    trace.add('population={}'.format(population))
    if population:
        variance = ssd / num_items
        trace.add('variance={}'.format(variance))
        return variance
    else:
        variance = ssd / (num_items - 1)
        sd = sqrt(variance)
        trace.add('sd={}'.format(sd))

    return sd


if __name__ == '__main__':
    unittest.main()


logtrace/__init__.py
# -*- coding: utf-8 -*-

# Copyright (c) 2017-2018, Paul Wolf.
# All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:

# 1. Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.

# 2. Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.

# 3. Neither the name of Yewleaf Ltd. nor the names of its contributors
# may be used to endorse or promote products derived from this software
# without specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# Original author: paul.wolf@yewleaf.com

import uuid
import time
import inspect
import logging

__version__ = '0.1.2'
__author__ = 'Paul Wolf'
__license__ = 'BSD'

__all__ = (
    'LogTrace',
)


def clean(s, chars):
    """Clean string s of characters in chars."""
    translation_table = dict.fromkeys(map(ord, chars), '_')
    return s.translate(translation_table)


def parse(s, delimiter):
    """
    https://stackoverflow.com/questions/18092354/python-split-string-without-splitting-escaped-character/18092547#18092547

    https://www.balabit.com/documents/syslog-ng-ose-latest-guides/en/syslog-ng-ose-guide-admin/html/concepts-message-ietfsyslog.html

    """
    pass


class LogTrace(object):
    def __init__(self,
                 logger=None,  # we'll emit output here
                 delimiter="; ",  # delimiter between parts
                 tag='',  # add a label to the log entry, non-unique
                 unique_id=False,  # create a uuid to identify the log?
                 level=None,  # default level if you want to use emit()
                 verbosity='v'):
        self.uid = None
        self.verbosity = verbosity
        self.data = {}
        if logger:
            self.logger = logger
        else:
            self.logger = logging.getLogger(__name__)
        self.delimiter = delimiter
        if level:
            self.level = level
        else:
            self.level = self.logger.getEffectiveLevel()

        self.event_log = []

        if tag:  # prepend log record tag
            if not tag[0] == '#':
                tag = '#' + tag
            self.event_log.append(tag)
        if unique_id:
            self.uid = uuid.uuid4()
        self.clean = False

        self.emit_funcs = {
            logging.DEBUG: self.logger.debug,
            logging.INFO: self.logger.info,
            logging.WARNING: self.logger.warning,
            logging.ERROR: self.logger.error,
            logging.CRITICAL: self.logger.critical,
        }

        self.start = time.time()

    def clear(self):
        self.event_log = []
        self.start = time.time()

    def get_uid(self):
        if not self.uid:
            self.uid = uuid.uuid4()
        return self.uid

    def set_uid(self, uid):
        """
        Accepts a string or uuid. Adds it to self.uid as a uuid.
        We want to not care which it is. Any string can be used for this.
        """
        if not (isinstance(uid, uuid.UUID) or isinstance(uid, str)):
            uid = str(uid)
        self.uid = uid

    def _add_message(self, msg, backup=1):
        """Append msg to logs only if log level is effective."""
        if not self.logger.getEffectiveLevel() >= self.logger.level:
            print("WARNING, not adding message, effective level: {}, level={}".format(self.logger.getEffectiveLevel(),
                                                                                      self.logger.level))
            return

        f = inspect.currentframe()
        for __ in range(backup):
            f = f.f_back
            i = inspect.getframeinfo(f)
        if self.verbosity == 'v':
            msg = "[{0}, {1:.4f}s] {2}".format(i.lineno, time.time() - self.start, msg)
        elif self.verbosity == 'vv':
            msg = "[{0}:{1}, {2:.4f}s] {3}".format(i.function,
                                                   i.lineno, time.time() - self.start, msg)
        elif self.verbosity == 'vvv':
            msg = "[{0}.{1}:{2}, {3:.4f}s] {4}".format(i.filename, i.function,
                                                       i.lineno, time.time() - self.start, msg)
        self.event_log.append(msg)

    def _add_data(self, data):
        """Update self.data with new data, be careful with key duplicates. """
        self.data.update(data)

    def add(self, msg=None, data=None, backup=2):
        if msg:
            self._add_message(msg, backup)
        if data:
            self._add_data(data)

    def build_message(self, delimiter):
        msg = delimiter.join(self.event_log)
        if self.clean:
            msg = msg.replace(delimiter, '_')

        if self.uid:
            msg = "uid={}{}{}".format(self.uid, self.delimiter, msg)
        return msg

    def emit_string(self, msg=None, delimiter=None, backup=2):
        """Return string of log record."""

        if not delimiter:
            delimiter = self.delimiter

        if msg:
            self.add(msg, backup=backup)

        return self.build_message(delimiter=delimiter)

    def emit(self, msg=None, delimiter=None, emit_func=None, backup=2, **extra):
        """Call emit function on self.logger which
        defaults to logger.debug

        or call emit_func which should be something like:

            log.emit(emit_func=logger.error)

        backup: how many frames to go back in the stack
        for the actually relevant calling frame

        If you want to pass any additional options into logger.debug etc
            functions - just use **extra params.
        """
        if not self.logger.getEffectiveLevel() >= self.logger.level:
            return

        if not delimiter:
            delimiter = self.delimiter

        if msg:
            self.add(msg, backup=backup)

        if not emit_func:
            emit_func = self.emit_funcs[self.level]

        emit_func(self.build_message(delimiter=delimiter), **extra)

    def emit_error(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.error, backup=3)

    def emit_debug(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.debug, backup=3)

    def emit_info(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.info, backup=3)

    def emit_warning(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.warning, backup=3)

    def emit_critical(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.critical, backup=3)


logs.py
# -*- coding: utf-8 -*-
import uuid
import time
import socket
import inspect
import logging
import warnings

# logger = logging.getLogger(__name__)

__all__ = (
    'LogTrace',
)

def clean(s, chars):
    """Clean string s of characters in chars."""
    translation_table = dict.fromkeys(map(ord, chars), '_')
    return s.translate(translation_table)    

def parse(s, delimiter):
    """
    https://stackoverflow.com/questions/18092354/python-split-string-without-splitting-escaped-character/18092547#18092547

    https://www.balabit.com/documents/syslog-ng-ose-latest-guides/en/syslog-ng-ose-guide-admin/html/concepts-message-ietfsyslog.html

    """
    pass
    
class LogTrace(object):
    def __init__(self,
                 logger=None,  # we'll emit output here
                 delimiter="; ",  # delimiter between parts
                 tag='',  # add a label to the log entry, non-unique
                 unique_id=False,  # create a uuid to identify the log?
                 level=logging.DEBUG):
        self.uid = None
        self.data = {}
        if logger:
            self.logger = logger
        else:
            self.logger = logging.getLogger("logtrace")
        self.delimiter = delimiter
        self.level = level
        self.tag = tag
        self.event_log = []
        if tag:  # prepend log record tag
            self.event_log.append(tag)
        if unique_id:
            self.uid = uuid.uuid4()

        self.start = time.time()

    def get_uid(self):
        if not self.uid:
            self.uid = uuid.uuid4()
        return self.uid

    def set_uid(self, uid):
        """
        Accepts a string. Adds it to self.uid as a uuid
        """
        if isinstance(uid, uuid.UUID):
            uid = uuid.UUID(uid)
        self.uid = uid

    def _add_message(self, msg, backup=1):
        """Append msg to logs only if log level is effective."""
        if not self.logger.getEffectiveLevel() >= self.level:
            return

        f = inspect.currentframe()
        for __ in range(backup):
            f = f.f_back
            i = inspect.getframeinfo(f)
        msg = "[{0}, {1:.4f}s] {2}".format(i.lineno, time.time() - self.start, msg)

        self.event_log.append(msg)

    def _add_data(self, data):
        """Update self.data with new data, be careful with key duplicates. """
        self.data.update(data)

    def add(self, msg=None, data=None, backup=1):
        if msg:
            self._add_message(msg, backup)
        if data:
            self._add_data(data)

    def build_message(self):
        msg = delimiter.join(self.event_log)
        if self.clean:
            msg = msg.replace(delimiter, '_')

        if self.uid:
            msg = "uid={}{}{}".format(self.uid, self.delimiter, msg)
        return msg

    def emit_string(self, msg=None, delimiter=None, backup=2):
        """Return string of log record."""

        if not delimiter:
            delimiter = self.delimiter

        if msg:
            self.add(msg, backup=backup)

        return self.build_message()

    def emit(self, msg=None, delimiter=None, emit_func=None, backup=2):
        """Call emit function on self.logger which 
        defaults to logger.debug

        or call emit_func which should be something like:

            log.emit(emit_func=logger.error)

        backup: how many frames to go back in the stack 
        for the actually relevant caller

        """
        if not delimiter:
            delimiter = self.delimiter

        if msg:
            self.add(msg, backup=backup)

        extra = {}

        if not emit_func:
            self.logger.debug(self.build_message(), extra=extra)
        else:
            emit_func(self.build(), extra=extra)

    def emit_error(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.error, backup=3)

    def emit_debug(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.debug, backup=3)

    def emit_info(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.info, backup=3)

    def emit_warning(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.warning, backup=3)

    def emit_critical(self, msg, delimiter=None):
        self.emit(msg, delimiter, emit_func=self.logger.critical, backup=3)



